codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocket13FrameDecoder.java

```diff
@@ -286,25 +284,29 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in) throws Exception {
             }
 
             state = State.MASKING_KEY;
-        case MASKING_KEY:
+        }
+        case MASKING_KEY: {
             if (frameMasked) {
                 if (in.readableBytes() < 4) {
                     return;
                 }
                 if (maskingKey == null) {
                     maskingKey = new byte[4];
                 }
-                in.readBytes(maskingKey);
+                in.copyInto(in.readerOffset(), maskingKey, 0, maskingKey.length);
+                in.skipReadable(maskingKey.length);
             }
             state = State.PAYLOAD;
-        case PAYLOAD:
+        }
+        case PAYLOAD: {
             if (in.readableBytes() < framePayloadLength) {
                 return;
             }
 
-            ByteBuf payloadBuffer = null;
+            Buffer payloadBuffer = null;
             try {
-                payloadBuffer = readBytes(ctx.alloc(), in, toFrameLength(framePayloadLength));
+                in.readSplit(0).close(); // Truncate off the frame header, so payload start at offset 0.
```

Comment:
this looks a bit odd ... Maybe there is a "better way" to do this ?

Reply:
The `*Frame` objects currently assume that they start on offset 0 in their respective buffers. We wouldn't need to do this if they instead relied on the read offset.

Reply:
Hmm we shouldn't expect this